lib/sequel/dataset/actions.rb

```diff
@@ -173,9 +173,9 @@ def each
     #   DB[:table].empty? # SELECT 1 AS one FROM table LIMIT 1
     #   # => false
     def empty?
-      cached_dataset(:_empty_ds) do
+      with_sql_first((cached_dataset(:_empty_ds) do
```

Comment:
We should leave the existing code as is for datasets not using custom SQL, because it prevents loading a large amount of data.  For datasets using custom SQL, maybe we should call `from_self`, either in this method, or in `single_value_ds` if this also affects other methods.  If in this method, maybe something like:

```diff
-        single_value_ds.unordered.select(EMPTY_SELECT)
+        (@opts[:sql] ? from_self : self).single_value_ds.unordered.select(EMPTY_SELECT)
```

Reply:
Thanks for the suggestion. I tried it (making the suggested change and reverting the lines above and below it), it fixes the custom SQL cases, but makes `DB.values([[nil, 1], [2, 3]]).empty?` incorrect again as it is not custom SQL. How can we also detect this case, and are there any other cases that need to be worked around?

My thought process when I came up with the code I shared was as follows, please correct my assumptions in the comments:
```ruby
# original code for reference, unchanged
def with_sql_single_value(sql)
  if r = with_sql_first(sql)
    r.each{|_, v| return v}
  end
end

# original code for reference, unchanged
def single_value!
  with_sql_single_value(select_sql)
end

# 1. original code,
# where dataset = (cached_dataset(...) {...}) since that is unchanged
def empty?
  dataset.single_value!.nil?
end

# 2. inline the call to single_value!
def empty?
  with_sql_single_value(dataset.select_sql).nil?
end

# 3. inline the call to with_sql_single_value
def empty?
  (if r = with_sql_first(dataset.select_sql)
    r.each{|_, v| break v}
  end).nil?
end

# 4. fix the bug: if there is a row we should return false,
# without checking if the first iterated column of the row is also not nil
def empty?
  (if r = with_sql_first(dataset.select_sql)
    r
  end).nil?
end

# 5. simplifying, assuming with_sql_first never returns false
def empty?
  with_sql_first(dataset.select_sql).nil?
end
```

Would you mind sharing how the code prevents loading a large amount of data? I'm not yet familiar enough with the library and was unaware of this case and how to trigger it. Moreover, I'm not sure which part of the above transformation would break this (loading a large amount of data) case specifically.